R/epidatacall.R

```diff
@@ -68,187 +58,219 @@ create_epidata_call <- function(endpoint, params, meta = NULL,
 #' @importFrom jsonlite fromJSON
 #' @return
-#' - For `fetch_classic`: the "classic" 3-element list format (see above)
+#' - For `fetch_classic`: a JSON-like list
 #'
 #' @export
-fetch_classic <- function(epidata_call, fields = NULL, disable_date_parsing = FALSE) {
+fetch_classic <- function(epidata_call, fields = NULL, disable_data_frame_parsing = TRUE) {
   stopifnot(inherits(epidata_call, "epidata_call"))
   stopifnot(is.null(fields) || is.character(fields))
-  stopifnot(is.logical(disable_date_parsing), length(disable_date_parsing) == 1)
+
   res <- request_impl(epidata_call, "classic", fields)
-  r <- httr::content(res, "text", encoding = "UTF-8")
-  if (httr::http_error(res)) {
-    # return message in case of error
-    return(list(result = 0, message = paste0("error: ", r), epidata = data.frame()))
+  if (res$status_code != 200) {
+    # 500, 429, 401 are possible
+    msg <- "fetch data from API"
+    if (http_type(res) == "text/html") {
+      # grab the error information out of the returned HTML document
+      msg <- paste(msg, ":", xml2::xml_text(xml2::xml_find_all(
+        xml2::read_html(content(res, 'text')),
+        "//p"
+      )))
+    }
+    httr::stop_for_status(res, task = msg)
   }
+  r <- httr::content(res, "text", encoding = "UTF-8")
 
-  m <- jsonlite::fromJSON(r)
-  if ("epidata" %in% names(m) && !(epidata_call$only_supports_classic)) {
-    m$epidata <- parse_data_frame(epidata_call, m$epidata, disable_date_parsing = disable_date_parsing)
+  r <- jsonlite::fromJSON(r, simplifyDataFrame = !disable_data_frame_parsing)
+  # success is 1, no results is -2
+  if (r$result != 1 && r$result != -2) {
+    rlang::abort(paste0("epidata error: ", r$message), "epidata_error")
+  } else {
+    r
   }
-  m
 }
 
 #' Fetches the data and returns the CSV text
-#' @rdname epidata_call
 #'
 #' @param epidata_call an instance of `epidata_call`
 #' @param fields filter fields
 #' @importFrom httr stop_for_status content
 #' @importFrom rlang abort
 #' @return
-#' - For `fetch_csv`: a string containing CSV text, with the `"epidata_csv"` class added
+#' - For `fetch_csv`: a string containing CSV text, with the
+#'   `"epidata_csv"` class added
 #'
-#' @export
-fetch_csv <- function(epidata_call, fields = NULL) {
+fetch_csv <- function(epidata_call, fields = NULL, disable_date_parsing = FALSE, disable_tibble_output = FALSE) {
   stopifnot(inherits(epidata_call, "epidata_call"))
   stopifnot(is.null(fields) || is.character(fields))
+
   if (epidata_call$only_supports_classic) {
-    rlang::abort("the endpoint only supports the classic message format, due to a non-standard behavior",
+    rlang::abort("This endpoint only supports the classic message format, due to a non-standard behavior. Use fetch_classic instead.",
       epidata_call = epidata_call,
       class = "only_supports_classic_format"
     )
   }
+
   res <- request_impl(epidata_call, "csv", fields)
   httr::stop_for_status(res)
```

Comment:
todo: same `stop_for_status` http error messaging stuff should be here as above.  Could be refactored into its own method.

Reply:
done